### PR TITLE
Token tracker substream added

### PR DESCRIPTION
**Pull request by anshulforyou**
Substream tracks the token transfer events of erc20, 721 and 1155 contracts

### CODE REVIEW

File: abis/erc20.json [hunk 1]
```diff
@@ -0,0 +1,230 @@
+[
+    {
+      "constant": true,
+      "inputs": [
```
Comment (0xsarvesh) [inline abis/erc20.json]: Can you align ABI file names. erc20 -> ERC20

File: consumer/go/main/main.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/streamingfast/substreams/client"
+	"github.com/streamingfast/substreams/manifest"
+	pbsubstreams "github.com/streamingfast/substreams/pb/sf/substreams/v1"
+	"google.golang.org/grpc"
+)
+
+var jwt_token string
+var endpoint string
+
+func main() {
```
Comment (0xsarvesh) [inline consumer/go/main/main.go]: Please remove this, we are not going to use this consumer.

Comment (0xsarvesh) [reply consumer/go/main/main.go]: Also remove the related go files.

File: consumer/python/main.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import os
+import grpc
+import sys
+
+from sf.substreams.v1 import substreams_pb2_grpc
+from sf.substreams.v1.substreams_pb2 import Request, STEP_IRREVERSIBLE
+from sf.substreams.v1.package_pb2 import Package
+
+jwt_token = os.getenv("SUBSTREAMS_API_TOKEN")
+if not jwt_token: raise Error("set SUBSTREAMS_API_TOKEN")
```
Comment (0xsarvesh) [inline consumer/python/main.py]: Remove python consumer logic. All the files.

File: proto/token-tracker.proto [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    repeated AirTransfer air_transfers = 1;
+    repeated AirERC1155TransferBatch air_transfer_batchs = 2;
+    repeated AirERC1155TransferSingle air_transfer_singles = 3;
+}
+
+message AirTransfer{
+    string hash = 1;
+    uint64 log_ordinal = 2;
+    string token_id = 3;
+    string from = 4;
```
Comment (0xsarvesh) [inline proto/token-tracker.proto]: Please add chainId in all the events.

File: src/abis/ERC721.rs [hunk 5]
```diff
@@ -0,0 +1,1500 @@
+    const INTERNAL_ERR: &'static str = "`ethabi_derive` internal error";
+    /// Contract's functions.
+    #[allow(dead_code)]
+    #[allow(unused_variables)]
+    pub mod functions {
+        use super::INTERNAL_ERR;
```
Comment (0xsarvesh) [inline src/abis/ERC721.rs]: Please check file name convention for all generated files,

File: src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+use substreams_ethereum::{pb::eth::v2 as eth, Event as EventTrait};
+
+#[substreams::handlers::map]
+fn map_transfers(blk: eth::Block) -> Result<AirTransfers, Error> {
+    let mut air_transfers = vec![];
+    let mut air_transfer_batchs = vec![];
+    let mut air_transfer_singles = vec![];
+    for log in blk.logs(){
+        if let Some(event) = abis::erc20::events::Transfer::match_and_decode(log){
+            log::info!("transfer event: {}", Hex(&event.to));
```
Comment (0xsarvesh) [inline src/lib.rs]: Please remove this log.

Comment (0xsarvesh) [reply src/lib.rs]: Lets add one log at the end, only txHash.

File: src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            // log::info!("ERC20 token: {}, {}, {}", Hex(log.log.clone().address), log.receipt.transaction.hash.clone().to_string(), log.ordinal());
+
+            let air_transfer: AirTransfer = AirTransfer{
+                amount: event.value.to_string(),
+                token_id: Hex(log.log.clone().address).to_string(),
```
Comment (0xsarvesh) [inline src/lib.rs]: Doesnt seem correct.

File: src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+                hash: Hex(log.receipt.transaction.hash.clone()).to_string(),
+                from: Hex(event.from).to_string(),
+                to: Hex(event.to).to_string(),
+                token_type: 0,
```
Comment (0xsarvesh) [inline src/lib.rs]: Can we use ENUM/String?

File: src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+                block_number: blk.clone().number,
```
Comment (0xsarvesh) [inline src/lib.rs]:
```suggestion
                block_number: blk.number,
```

File: proto/token-tracker.proto [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    string hash = 1;
+    uint64 log_ordinal = 2;
+    string token_id = 3;
+    string operator = 4;
+    string from = 5;
+    string to = 6;
+    string amount = 7;
+    AirTokenStandard token_type = 8;
+    uint64 block_number = 9;
+    string block_timestamp = 10;
```
Comment (0xsarvesh) [inline proto/token-tracker.proto]: timestamp should not be string, should be uint64

File: proto/token-tracker.proto [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+message AirTransfers{
+    repeated AirTransfer air_transfers = 1;
+    repeated AirERC1155TransferBatch air_transfer_batchs = 2;
+    repeated AirERC1155TransferSingle air_transfer_singles = 3;
+}
+
+message AirTransfer{
+    string hash = 1;
```
Comment (0xsarvesh) [inline proto/token-tracker.proto]: tokenAddress is missing in all three.

File: src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+                .to_string(),
+                token_ids: event.ids.iter().map(|id| Hex(id.clone()).0.to_string()).collect(),
+                log_ordinal: log.ordinal(),
+                ..Default::default()
+            };
+            air_transfer_batchs.push(air_transfer_batch);
+        }
+        if let Some(event) = abis::erc1155::events::TransferSingle::match_and_decode(log){
+            log::info!("transfer event: {}", Hex(&event.to));
+            let air_transfer_single: AirErc1155TransferSingle = AirErc1155TransferSingle{
```
Comment (0xsarvesh) [inline src/lib.rs]: please rename this to erc1155_transfer_single

File: src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                .to_string(),
+                log_ordinal: log.ordinal(),
+                ..Default::default()
+            };
+            air_transfers.push(air_transfer);
+        }
+        if let Some(event) = abis::erc1155::events::TransferBatch::match_and_decode(log){
+            log::info!("transfer event: {}", Hex(&event.to));
+            // log::info!("transfer values: {}", &event.values);
+            let air_transfer_batch: AirErc1155TransferBatch = AirErc1155TransferBatch{
```
Comment (0xsarvesh) [inline src/lib.rs]: please rename this to erc1155_transfer_batch

File: src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+                log_ordinal: log.ordinal(),
+                ..Default::default()
+            };
+    
+            air_transfers.push(air_transfer);
+        }
+        if let Some(event) = abis::erc721::events::Transfer::match_and_decode(log){
+            log::info!("transfer event: {}", Hex(&event.to));
+            // log::info!("ERC721 token: {}, {}", log.receipt.transaction.hash.clone().to_string(), log.ordinal());
+            let air_transfer: AirTransfer = AirTransfer{
```
Comment (0xsarvesh) [inline src/lib.rs]: please rename this to erc20_transfer

File: src/rpc.rs [hunk 15]
```diff
@@ -0,0 +1,24 @@
+// use bigdecimal::BigDecimal;
+// use num_bigint::BigInt;
+// use substreams::log;
+// use substreams_ethereum::pb::eth as ethpb;
+
+// pub fn get_wallet_balance(wallet_address: &String) -> BigInt{
+//     let rpc_calls: ethpb::rpc::RpcCalls = create_wallet_balance_call(&hex::decode(wallet_address).unwrap());
```
Comment (0xsarvesh) [inline src/rpc.rs]: Delete this file.

File: substreams.yaml [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+binaries:
+  default:
+    type: wasm/rust-v1
+    file: target/wasm32-unknown-unknown/release/all_token_tracker_substream.wasm
+
+modules:
+  - name: map_transfers
+    kind: map
+    initialBlock: 12369621
```
Comment (0xsarvesh) [inline substreams.yaml]: Change this to block 0

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]

**Comment by 0xsarvesh** [comment]
@anshulforyou  Please add a transfer event for base token transfer.

File: proto/token-tracker.proto [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+enum AirTokenStandard{
+    UNKNOWN = 0;
+    ERC20 = 1;
+    ERC721 = 2;
+    ERC1155 = 3;
+}
+
+message AirTransfers{
+    repeated AirTransfer air_transfers = 1;
+    repeated AirERC1155TransferBatch air_transfer_batchs = 2;
```
Comment (0xsarvesh) [inline proto/token-tracker.proto]:
```suggestion
    repeated AirERC1155TransferBatch air_erc1155_transfer_batchs = 2;
```

File: proto/token-tracker.proto (hunk 11 above)
Comment (anshulforyou) [reply proto/token-tracker.proto]: Should I keep separate token address and token id for erc1155 as well??

File: src/lib.rs (hunk 7 above)
Comment (anshulforyou) [reply src/lib.rs]: Which line exactly?

**Comment by 0xsarvesh** [comment]
can you move overall project to specific folder. We are going to use this one repo for many substreams.

File: abis/erc20.json (hunk 1 above)
Comment (anshulforyou) [reply abis/erc20.json]: We have to keep everything snake case else rust starts showing warning for it

Comment (anshulforyou) [reply abis/erc20.json]: I have renamed all abi files to snake case "erc20"

File: consumer/go/main/main.go (hunk 2 above)
Comment (anshulforyou) [reply consumer/go/main/main.go]: Done

File: consumer/python/main.py (hunk 3 above)
Comment (anshulforyou) [reply consumer/python/main.py]: Done

File: proto/token-tracker.proto (hunk 4 above)
Comment (anshulforyou) [reply proto/token-tracker.proto]: Done, for now I am entering it manually. Will make rpc call for it in future

File: src/abis/ERC721.rs (hunk 5 above)
Comment (anshulforyou) [reply src/abis/ERC721.rs]: Done

File: src/lib.rs (hunk 6 above)
Comment (anshulforyou) [reply src/lib.rs]: Done, only tx hash is being logged now

File: src/lib.rs (hunk 7 above)
Comment (anshulforyou) [reply src/lib.rs]: I have changed token_id to token_address for this

File: src/lib.rs (hunk 8 above)
Comment (anshulforyou) [reply src/lib.rs]: Yeah it is already string

File: src/lib.rs (hunk 9 above)
Comment (anshulforyou) [reply src/lib.rs]: Done

File: proto/token-tracker.proto (hunk 10 above)
Comment (anshulforyou) [reply proto/token-tracker.proto]: Done

File: src/lib.rs (hunk 12 above)
Comment (anshulforyou) [reply src/lib.rs]: Done

File: src/lib.rs (hunk 13 above)
Comment (anshulforyou) [reply src/lib.rs]: Done

File: src/lib.rs (hunk 14 above)
Comment (anshulforyou) [reply src/lib.rs]: Done

File: src/rpc.rs (hunk 15 above)
Comment (anshulforyou) [reply src/rpc.rs]: Done

File: substreams.yaml (hunk 16 above)
Comment (anshulforyou) [reply substreams.yaml]: Done

File: proto/token-tracker.proto (hunk 17 above)
Comment (anshulforyou) [reply proto/token-tracker.proto]: Done

File: proto/token-tracker.proto (hunk 11 above)
Comment (stevepro-lab) [reply proto/token-tracker.proto]: Yes tokenAddress is contract address & tokenId id nft id

**Pull request merged**